pingen/models/pingen.py

```diff
@@ -0,0 +1,298 @@
+# Author: Guewen Baconnier
+# Copyright 2012-2023 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+import json
+import logging
+from datetime import datetime
+from urllib.parse import urljoin
+
+import pytz
+import requests
+from dateutil import parser
+from oauthlib.oauth2 import BackendApplicationClient
+from requests_oauthlib import OAuth2Session
+
+_logger = logging.getLogger(__name__)
+
+
+def pingen_datetime_to_utc(dt):
+    """Convert a date/time used by pingen.com to UTC timezone
+
+    :param dt: pingen date/time iso string (as received from the API)
+               to convert to UTC
+    :return: TZ naive datetime in the UTC timezone
+    """
+    utc = pytz.utc
+    localized_dt = parser.parse(dt)
+    return localized_dt.astimezone(utc).replace(tzinfo=None)
+
+
+class PingenException(RuntimeError):
+    """There was an ambiguous exception that occurred while handling your
+    request."""
+
+
+class APIError(PingenException):
+    """An Error occured with the pingen API"""
+
+
+class Pingen(object):
+    """Interface to the pingen.com API"""
+
+    def __init__(self, clientid, secretid, organization, staging=True):
+        self.clientid = clientid
+        self.secretid = secretid
+        self.organization = organization
+        self.staging = staging
+        self._session = None
+        self._init_token_registry()
+        super(Pingen, self).__init__()
+
+    @property
+    def api_url(self):
+        if self.staging:
+            return "https://api-staging.v2.pingen.com"
+        return "https://api.v2.pingen.com"
```

Comment:
The urls with ".v2." will be discontinued in April 2024

Reply:
as there is card for that, we will followup in another PR, so lets try to merge this one first.